<?xml version="1.0"?>
<launch>
	<arg name="master_name" default="$(arg robot_name)"/>
	<arg name="robot_name" default="jethexa" />
	<arg name="tf_prefix" default="jethexa" />

	<!--whether to use simulation-->
	<arg name="sim" default="false" />

	<!--whether to use depth camera to simulate Lidar-->
	<arg name="use_depth_camera" default="false"/>
	<arg name="depth_camera_name" default="astra_cam" />

	<arg name="map_topic" default="/$(arg master_name)/map"/>
	<arg name="scan_topic" default="scan/raw"/>
	<arg name="odom_topic" default="odom/filtered"/>
	<arg name="cmd_vel_topic" default="jethexa_controller/cmd_vel"/>
	<arg name="odom_frame_id" default="$(arg tf_prefix)/odom"/>
	<arg name="base_frame_id" default="$(arg tf_prefix)/base_link"/>
	<arg name="map_frame_id" default="map"/>
	<arg name="move_base_result" default="move_base/result" />
	<arg name="clicked_point" default="clicked_point" />

	<arg name="open_rviz" default="true"/>

	<group ns="/">
		<!-- URDF description of robot -->
		<include file="$(find jethexa_description)/launch/description.launch">
			<arg name="robot_name" value="$(arg robot_name)" />
			<arg name="tf_prefix" value="$(arg tf_prefix)" />
		</include>

		<!--whether to use simulation-->
		<group unless="$(arg sim)">
			<!-- camera -->
			<include file="$(find jethexa_peripherals)/launch/camera_ros_app.launch">
				<arg name="tf_prefix" value="$(arg tf_prefix)" />
			</include>

			<!-- Lidar -->
			<include unless="$(arg use_depth_camera)" file="$(find jethexa_peripherals)/launch/lidar.launch">
				<arg name="tf_prefix" value="$(arg tf_prefix)" />
				<arg name="scan_topic" value="$(arg scan_topic)" />
			</include>

			<!-- robot motion control -->
			<include file="$(find jethexa_controller)/launch/jethexa_controller_ros_app.launch">
				<arg name="robot_name" value="$(arg robot_name)" />
				<arg name="tf_prefix" value="$(arg tf_prefix)" />
				<arg name="tf_enable" value="false" />
				<arg name="odom_enable" value="false" />
			</include>

			<!-- posture sensor -->
			<!--
			<include file="$(find jethexa_peripherals)/launch/imu.launch">
					<arg name="tf_prefix" value="$(arg tf_prefix)" />
					<arg name="freq" value="50" />
			</include>
			-->

			<!-- Lidar simulates odometer -->
			<include file="$(find jethexa_slam)/launch/include/rf2o_laser_odometry.launch">
				<arg name="scan_topic" value="$(arg scan_topic)" />
				<arg name="odom_topic" value="odom/laser" />
				<arg name="odom_frame_id" value="$(arg odom_frame_id)" />
				<arg name="base_frame_id" value="$(arg base_frame_id)" />
				<arg name="laser_frame_id" value="$(arg tf_prefix)/laser_link" />
			</include>

			<!-- odometer integrates filter -->
			<include file="$(find jethexa_slam)/launch/include/jethexa_ekf.launch">
				<arg name="tf_prefix" value="$(arg tf_prefix)" />
				<arg name="robot_name" value="$(arg robot_name)" />
			</include>
		</group>

		<!-- start the AMCL Adaptive Monte Carlo Localization algorithm package -->
		<include file="$(find jethexa_navigation)/launch/include/amcl.launch">
			<arg name="scan_topic" value="$(arg scan_topic)"/>
			<arg name="map_topic" value="$(arg map_topic)"/>
			<arg name="map_frame_id" value="$(arg map_frame_id)"/>
			<arg name="odom_frame_id" value="$(arg odom_frame_id)"/>
			<arg name="base_frame_id" value="$(arg base_frame_id)"/>
			<arg name="use_map_topic" value="true"/>
		</include>

		<!-- start the path planning algorithm package -->
		<include file="$(find jethexa_navigation)/launch/include/move_base.launch">
			<arg name="cmd_vel_topic" value="$(arg cmd_vel_topic)" />

			<arg name="global_costmap_map_topic" value="$(arg map_topic)"/>
			<arg name="global_costmap_sensor_frame" value="$(arg base_frame_id)"/>
			<arg name="global_costmap_sensor_topic" value="$(arg scan_topic)"/>
			<arg name="global_costmap_global_frame" value="$(arg map_frame_id)"/>
			<arg name="global_costmap_robot_base_frame" value="$(arg base_frame_id)"/>

			<arg name="local_costmap_map_topic" value="$(arg map_topic)"/>
			<arg name="local_costmap_sensor_frame" value="$(arg base_frame_id)"/>
			<arg name="local_costmap_sensor_topic" value="$(arg scan_topic)"/>
			<arg name="local_costmap_global_frame" value="$(arg odom_frame_id)"/>
			<arg name="local_costmap_robot_base_frame" value="$(arg base_frame_id)"/>

			<arg name="teb_odom_topic" value="$(arg odom_topic)"/>
			<arg name="teb_map_frame" value="$(arg odom_frame_id)"/>
		</include>

		<!-- multi-point navigation -->
		<node pkg="jethexa_navigation" type="publish_point.py" name="publish_point" output="screen">
			<param name="map_frame" value="$(arg map_frame_id)" />
			<param name="clicked_point" value="$(arg clicked_point)" />
			<param name="move_base_result" value="$(arg move_base_result)" />
		</node>
	</group>
</launch>
